### DISCUSSION TITLE
runtime environment variables in custom server (shared hosting after building on local environment)

### DISCUSSION THREAD

**Discussion by robiulhr** [vercel/next.js#69552] [Help]
### Summary

I am using shared hosting to host and can't build the project on the server. I build the project on the local environment and then upload it to the shared hosting. Since it's already built, it has nothing to do with the environment variable on my hosting environment. But I need to use the public env variables in the project. 

My project is running by the next js custom server, I have tried a couple of ways and finally reached this code from v0.dev 

```js
    const nonce = crypto.randomBytes(16).toString('base64')

// Create a wrapper for the response object
      const wrappedRes = new Proxy(res, {
        get: function (target, prop) {
          if (prop === 'writeHead') {
            return function (...args) {
              // Inject environment variables before writing headers
              const script = `<script nonce="${nonce}">
              window.ENV = {
                NEXT_PUBLIC_WEB_PUSH_PUBLIC_KEY: "${process.env.NEXT_PUBLIC_WEB_PUSH_PUBLIC_KEY}"
              }
            </script>`

              // Set Content-Security-Policy header with nonce
              const csp = `script-src 'self' 'nonce-${nonce}' 'unsafe-inline'; style-src 'self' 'unsafe-inline'`
              args[1] = { ...args[1], 'Content-Security-Policy': csp }

              target.write(script)
              return target.writeHead.apply(target, args)
            }
          }
          return target[prop]
        }
      })
``` 

But I have not been able to make it work.

**Comment by Juneezee** [comment]
The documentation has already explained this, see 

1. https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser

   > After being built, your app will no longer respond to changes to these environment variables. ... If you need access to runtime environment values, you'll have to setup your own API to provide them to the client (either on demand or during initialization).



2. https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#runtime-environment-variables

   > However, these public environment variables will be inlined into the JavaScript bundle during `next build`.
